### PR TITLE
AttitudePositionEstimatorEKF fixes

### PR DESCRIPTION
**Pull request by Zefz**
This addresses some issues with using the EKF algorithm. First of all the estimator did not publish altitude position estimates unless we had previously at least 1 good GPS fix. This is not necessary because the Barometer can give us relative altitude estimates without a GPS fix (although no absolute AMSL estimate).

Without this, the system would not be able to run the Land Detector algorithm or use AltHold mode without at least one GPS fix.

Secondly it addresses an issue where the AMSL reference was incorrectly initialized without GPS. Previously the alt ref would begin at zero and slowly converge to the desired ref (which could give a ~80 meter error converging towards 0 the first 20 or so seconds).

This PR requires some rigorous testing.

### CODE REVIEW

**Comment by Zefz** [comment]
Now also includes fix for the new secondary sensor fallback logic. Fixes issue where the system would constantly toggle between magnetometer 0 and 1 and IMU/GYRO switch from 0 to 1.

**Comment by LorenzMeier** [comment]
It looks like this contains #1746?

**Comment by Zefz** [comment]
This has been flight-tested and verified using QuadCopter. Works like a charm.

**Comment by Zefz** [comment]
I've pushed my GPS debug code by mistake. These will be removed in the future again.

**Comment by Zefz** [comment]
This has now been flight-tested and verified that solves issue #1777 along with some other bugs that the EKF had when flying Multicopters.

**Comment by LorenzMeier** [comment]
@sjwilks Could you give this some fixed-wing validation flight testing? Looks like a lot of progress.

File: src/drivers/gps/gps.cpp [hunk 1]
```diff
@@ -664,6 +675,14 @@ gps_main(int argc, char *argv[])
 		gps::start(device_name, fake_gps, enable_sat_info);
 	}
 
+	if (!strcmp(argv[1], "fake")) {
```
Comment (LorenzMeier) [inline src/drivers/gps/gps.cpp]: Why can't it be started here with just -f?

**Comment by LorenzMeier** [comment]
This has been superseded by #1794.

**Pull request closed without merging**